File: pptx-service/renderers/kpi_board.py
import logging

from builder import add_bg, add_textbox, add_rect, add_line, DELTA_COLORS
from theme import ThemeColors

logger = logging.getLogger(__name__)


def render(slide, data: dict, t: ThemeColors) -> None:
    add_bg(slide, t.bg)
    heading = data.get("heading", "")
    period = data.get("period", "")
    kpis = data.get("kpis", [])
    takeaway = data.get("takeaway", "")

    add_textbox(slide, 0.83, 0.3, 9.0, 0.8, heading,
                t.font_display, 40, t.text, bold=True)
    if period:
        add_textbox(slide, 10.0, 0.3, 3.0, 0.8, period,
                    t.font_body, 20, t.muted, align="right")

    if len(kpis) > 6:
        logger.warning("kpi-board: %d KPIs provided, displaying first 6", len(kpis))
    kpis = kpis[:6]
    n = len(kpis)
    if n == 0:
        return
    cols = 3 if n > 3 else n
    rows = (n + cols - 1) // cols
    cell_w = 12.0 / cols
    cell_h = (5.0 if takeaway else 5.5) / rows

    for i, kpi in enumerate(kpis):
        col = i % cols
        row = i // cols
        x = 0.67 + col * cell_w
        y = 1.3 + row * cell_h

        add_rect(slide, x, y, cell_w - 0.2, cell_h - 0.1,
                 fill_hex=t.paper, line_hex=t.border, line_width_pt=0.5)

        add_textbox(slide, x + 0.15, y + 0.12, cell_w - 0.5, 0.4,
                    kpi.get("label", ""), t.font_body, 18, t.muted)
        add_textbox(slide, x + 0.15, y + 0.5, cell_w - 0.5, 1.0,
                    kpi.get("value", ""), t.font_display, 44, t.text, bold=True)
        delta = kpi.get("delta", "")
        if delta:
            tone = kpi.get("deltaTone", "flat")
            add_textbox(slide, x + 0.15, y + 1.5, cell_w - 0.5, 0.35,
                        delta, t.font_body, 18,
                        DELTA_COLORS.get(tone, t.muted), bold=True)
        hint = kpi.get("hint", "")
        if hint:
            add_textbox(slide, x + 0.15, y + 1.9, cell_w - 0.5, 0.4,
                        hint, t.font_body, 15, t.muted, italic=True)

    if takeaway:
        add_line(slide, 0.83, 6.8, 12.5, 6.8, t.border, 0.5)
        add_textbox(slide, 0.83, 6.9, 11.5, 0.45, takeaway,
                    t.font_body, 20, t.muted, italic=True)

File: lib/pptx/slides/kpi-board.ts
import type { KpiBoardSlide } from '@/lib/types';
import {
  px, py, SLIDE_W,
  setBackground, addRect, addTextBox, addPageNumber,
  type PptxTheme,
} from '../helpers';
import type PptxGenJS from 'pptxgenjs';
type PSlide = ReturnType<InstanceType<typeof PptxGenJS>['addSlide']>;

const DELTA_POS = '22B800';
const DELTA_NEG = 'D93025';

export function render(slide: PSlide, data: KpiBoardSlide, t: PptxTheme, n: number, total: number): void {
  setBackground(slide, '#' + t.bg);

  addTextBox(slide, data.heading, {
    x: t.paddingIn, y: py(60), w: SLIDE_W - t.paddingIn - px(300),
    fontSize: t.sectionSize, color: '#' + t.text, fontFace: t.fontDisplay, bold: true,
  });
  addTextBox(slide, data.period, {
    x: SLIDE_W - t.paddingIn - px(300), y: py(70), w: px(290),
    fontSize: t.captionSize, color: '#' + t.muted, fontFace: t.fontBody, align: 'right',
  });

  const kpis = data.kpis.slice(0, 6);
  const cols = kpis.length <= 4 ? 4 : Math.min(kpis.length, 3);
  const rows = Math.ceil(kpis.length / cols);
  const cardW = (SLIDE_W - 2 * t.paddingIn - (cols - 1) * px(20)) / cols;
  const cardH = (py(700) - (rows - 1) * py(20)) / rows;
  const startY = py(180);

  kpis.forEach((kpi, i) => {
    const col = i % cols;
    const row = Math.floor(i / cols);
    const x = t.paddingIn + col * (cardW + px(20));
    const y = startY + row * (cardH + py(20));

    addRect(slide, x, y, cardW, cardH, '#' + t.rule, { transparency: 90 });
    addTextBox(slide, kpi.value, {
      x: x + px(16), y: y + py(16), w: cardW - px(32),
      fontSize: Math.round(t.heroSize * 0.7), color: '#' + t.text, fontFace: t.fontDisplay, bold: true,
    });

    if (kpi.delta) {
      const dc = kpi.deltaTone === 'pos' ? DELTA_POS : kpi.deltaTone === 'neg' ? DELTA_NEG : t.muted;
      addTextBox(slide, kpi.delta, {
        x: x + px(16), y: y + py(130), w: cardW - px(32),
        fontSize: t.captionSize, color: '#' + dc, fontFace: t.fontBody,
      });
    }

    addTextBox(slide, kpi.label, {
      x: x + px(16), y: y + cardH - py(80), w: cardW - px(32),
      fontSize: t.captionSize, color: '#' + t.muted, fontFace: t.fontBody,
    });
  });

  if (data.takeaway) {
    addRect(slide, t.paddingIn, py(900), SLIDE_W - 2 * t.paddingIn, py(60), '#' + t.accent, { transparency: 80 });
    addTextBox(slide, data.takeaway, {
      x: t.paddingIn + px(16), y: py(905), w: SLIDE_W - 2 * t.paddingIn - px(32),
      fontSize: t.bodySize, color: '#' + t.text, fontFace: t.fontBody,
    });
  }

  addPageNumber(slide, n, total, t);
}
